File: targets/java/runtime/org/transscript/compiler/std/NumExtern.java
/* Copyright (c) 2016 IBM Corporation. */

package org.transscript.compiler.std;

import static org.transscript.runtime.StringTerm.stringTerm;
import static org.transscript.runtime.DoubleTerm.doubleTerm;
import static org.transscript.runtime.Term.force;

import org.transscript.compiler.std.Core.Bool;
import org.transscript.runtime.Context;
import org.transscript.runtime.DoubleTerm;
import org.transscript.runtime.StringTerm;

public class NumExtern
{

	final public static DoubleTerm Plus(Context context, DoubleTerm left, DoubleTerm right)
	{
		DoubleTerm eleft = left.eval(context);
		DoubleTerm eright = right.eval(context);

		DoubleTerm r = DoubleTerm.doubleTerm(eleft.unbox() + eright.unbox());
		eleft.release();
		eright.release();
		return r;
	}

	public static Bool GreaterThan(Context context, DoubleTerm left, DoubleTerm right)
	{
		DoubleTerm eleft = left.eval(context);
		DoubleTerm eright = right.eval(context);

		Bool r = eleft.unbox() > eright.unbox() ? Core.TRUE(context) : Core.FALSE(context);
		eleft.release();
		eright.release();
		return r;
	}

	public static DoubleTerm Minus(Context context, DoubleTerm left, DoubleTerm right)
	{

		DoubleTerm eleft = left.eval(context);
		DoubleTerm eright = right.eval(context);

		DoubleTerm r = DoubleTerm.doubleTerm(eleft.unbox() - eright.unbox());
		eleft.release();
		eright.release();
		return r;
	}

	public static Bool LessThan(Context context, DoubleTerm left, DoubleTerm right)
	{
		DoubleTerm eleft = left.eval(context);
		DoubleTerm eright = right.eval(context);

		Bool r = eleft.unbox() < eright.unbox() ? Core.TRUE(context) : Core.FALSE(context);
		eleft.release();
		eright.release();
		return r;
	}
	

	/**
	 * 
	 * @param context
	 * @param num
	 * @return
	 */
	public static StringTerm FormatNumber(Context context, DoubleTerm num)
	{
		DoubleTerm evnum = force(context, num);

		// FIXME
		StringTerm result = stringTerm(Integer.toString((int) evnum.unbox()));
		evnum.release();

		return result;
	}

	public static StringTerm FormatDecimal(Context context, DoubleTerm value_1396)
	{
		// TODO Auto-generated method stub
		return null;
	}

	public static StringTerm FormatInteger(Context context, DoubleTerm num)
	{
		StringTerm result = stringTerm(Integer.toString((int) num.unbox()));
		num.release();
		return result;
	}

	public static DoubleTerm Hex(Context context, StringTerm value_1242)
	{
		// TODO Auto-generated method stub
		return null;
	}

	public static DoubleTerm ToInteger(Context context, StringTerm string)
	{
		DoubleTerm d = doubleTerm(Integer.parseInt(string.unbox()));
		string.release();
		return d;
	}

	public static DoubleTerm Times(Context context, DoubleTerm value_16555, DoubleTerm value_16556)
	{
		// FIXME
		return null;
	}

	public static DoubleTerm Divide(Context context, DoubleTerm value_16564, DoubleTerm value_16565)
	{
		// TODO Auto-generated method stub
		return null;
	}

	public static DoubleTerm Modulo(Context context, DoubleTerm value_16574, DoubleTerm value_16575)
	{
		// TODO Auto-generated method stub
		return null;
	}

	public static DoubleTerm ToDecimal(Context context, StringTerm value_16572)
	{
		// TODO Auto-generated method stub
		return null;
	}

}

File: targets/cpp/std/src/std/num-extern.cpp
// Copyright (c) 2016 IBM Corporation.
#include "num-extern.h"

#include <cstdio>
#include "std/core.h"

using namespace tosca;

// integer max and min definitions

/* @brief Internal number to represent integer max based on double's
          precision (mantissa is 53 bits). Note that integer is 
          language specific. In C++ the integer primitive is long 
          long (64 bits). In Java the integer primitive is int 
          (32 bits). */
#define INTEGER_MAX_DBL_PRECISION (1LL<<53)

/* @brief Internal number to represent integer min based on double's
          precision (mantissa is 53 bits). Note that integer is 
          language specific. In C++ the integer primitive is long 
          long (64 bits). In Java the integer primitive is int 
          (32 bits). */
#define INTEGER_MIN_DBL_PRECISION ((1LL<<53) * -1)

DoubleTerm& GetDefine_INTEGER_MAX_DBL_PRECISION(Context& ctx)
{
  return newDoubleTerm(ctx, INTEGER_MAX_DBL_PRECISION);
}

DoubleTerm& GetDefine_INTEGER_MIN_DBL_PRECISION(Context& ctx)
{
  return newDoubleTerm(ctx, INTEGER_MIN_DBL_PRECISION);
}

DoubleTerm& ClipToMaxOrMin(Context& ctx, DoubleTerm& dTerm)
{
  DoubleTerm result;
  double d = dTerm.Unbox();
  if (d > INTEGER_MAX_DBL_PRECISION)
  {
    dTerm.Release();
    return GetDefine_INTEGER_MAX_DBL_PRECISION(ctx);
  }
  else if (d < INTEGER_MIN_DBL_PRECISION)
  {
    dTerm.Release();
    return GetDefine_INTEGER_MIN_DBL_PRECISION(ctx);
  }
  else
  {
    dTerm.Release();
    return newDoubleTerm(ctx, d);
  }
}

Bool& IsWithinIntegerPrecision(Context& ctx, DoubleTerm& dTerm)
{
  double d = dTerm.Unbox();
  if (d > INTEGER_MAX_DBL_PRECISION ||
      d < INTEGER_MIN_DBL_PRECISION)
  {
    dTerm.Release();
    return newFALSE(ctx);
  }
  else
  {
    dTerm.Release();
    return newTRUE(ctx);
  }
}

DoubleTerm& Plus(Context& ctx, DoubleTerm& left, DoubleTerm& right)
{
    DoubleTerm& result = newDoubleTerm(ctx, left.Unbox() + right.Unbox());
    left.Release();
    right.Release();
    return result;
}

DoubleTerm& Minus(Context& ctx, DoubleTerm& left, DoubleTerm& right)
{
    DoubleTerm& result = newDoubleTerm(ctx, left.Unbox() - right.Unbox());
    left.Release();
    right.Release();
    return result;
}

DoubleTerm& Times(Context& ctx, DoubleTerm& left, DoubleTerm& right)
{
    DoubleTerm& result = newDoubleTerm(ctx, left.Unbox() * right.Unbox());
    left.Release();
    right.Release();
    return result;
}

DoubleTerm& Divide(Context& ctx, DoubleTerm& left, DoubleTerm& right)
{
    double uleft = left.Unbox();
    double uright = right.Unbox();
    DoubleTerm& result = newDoubleTerm(ctx, uleft / uright);
    left.Release();
    right.Release();
    return result;
}

DoubleTerm& Modulo(Context& ctx, DoubleTerm& left, DoubleTerm& right)
{
    long long uleft = (long long) left.Unbox();
    long long uright = (long long) right.Unbox();
    DoubleTerm& result = newDoubleTerm(ctx, uleft % uright);
    left.Release();
    right.Release();
    return result;
}


Bool& LessThan(Context& ctx, DoubleTerm& left, DoubleTerm& right)
{
    double uleft = left.Unbox();
    double uright = right.Unbox();
    Bool& result = uleft < uright ? newTRUE(ctx) : newFALSE(ctx);
    left.Release();
    right.Release();
    return result;
}

Bool& GreaterThan(Context& ctx, DoubleTerm& left, DoubleTerm& right)
{
    double uleft = left.Unbox();
    double uright = right.Unbox();
    Bool& result = uleft > uright ? newTRUE(ctx) : newFALSE(ctx);
    left.Release();
    right.Release();
    return result;
}

StringTerm& FormatNumber(Context& ctx, DoubleTerm& num)
{
    double unum = num.Unbox();
    num.Release();
    char* str = (char*) alloca((size_t) 32);
    snprintf(str, (size_t) 31, "%G", unum);
    return newStringTerm(ctx, str);
}

StringTerm& FormatInteger(Context& ctx, DoubleTerm& num)
{
    DoubleTerm& unumTerm = ClipToMaxOrMin(ctx, num);
    long long unum = (long long) unumTerm.Unbox();
    unumTerm.Release();
    char* str = (char*) alloca((size_t) 32);
    snprintf(str, (size_t) 31, "%lld", unum);
    return newStringTerm(ctx, str);
}

StringTerm& FormatDecimal(Context& ctx, DoubleTerm& num)
{
    double unum = num.Unbox();
    num.Release();
    char* str = (char*) alloca((size_t) 32);
    snprintf(str, (size_t) 31, "%f", unum);
    return newStringTerm(ctx, str);
}

tosca::DoubleTerm& ToDecimal(tosca::Context& ctx, tosca::StringTerm& str)
{
    double num;
    if (std::sscanf(str.Unbox().c_str(), "%lg", &num) <= 0)
    {
    	std::string s("Illegally formatted decimal: ");
    	s += str.Unbox().c_str();
        throw new std::runtime_error(s);
    }
    str.Release();
    return newDoubleTerm(ctx, num);
}


DoubleTerm& Hex(Context& ctx, StringTerm& hex)
{
    long long num;
    if (std::sscanf(hex.Unbox().c_str(), "%llX", &num) <= 0)
    {
    	std::string s("Illegally formatted hex: ");
    	s += hex.Unbox().c_str();

        throw new std::runtime_error(s);
    }
    hex.Release();
    return newDoubleTerm(ctx, num);
}
